### PR TITLE
Simplify and centralize the processing of internal states for TreeSelectControl

### PR DESCRIPTION
**Pull request by eason9487**
### Changes proposed in this Pull Request:

To simplify and centralize the processing of updating expanded, checked, and partially checked states, this PR:
- Replace `cloneDeep` with array `reduce` when generating filtered options.
- Calculate the expanded, checked, and partially checked states of options in the main component.
- Move the processing of updating the expanded state to the main component.
- Simplify the handling of updating the state of selected values.

### Detailed test instructions:

1. Go to the Edit free listings or the local storybook.
2. Test if the states and functionality of checking/unchecking options work correctly.
3. Test if the states and functionality of expanding/collapsing options work correctly.

### Changelog entry

> Tweak - Simplify and centralize the processing of internal states for the TreeSelectControl component.

### CODE REVIEW

File: js/src/components/tree-select-control/index.js [hunk 1]
```diff
@@ -92,14 +110,16 @@ const TreeSelectControl = ( {
...
 
 	const dropdownRef = useRef();
 	const onDropdownVisibilityChangeRef = useRef();
 	onDropdownVisibilityChangeRef.current = onDropdownVisibilityChange;
 
 	// We will save in a REF previous search filter queries to avoid re-query the tree and save performance
 	const filteredOptionsCache = useRef( {} );
+	const cacheStatesRef = useRef( {} );
+	cacheStatesRef.current.expandedValues = nodesExpanded;
+	cacheStatesRef.current.selectedValues = value;
```
Comment (puntope) [inline js/src/components/tree-select-control/index.js]: 💅  WDTY of saving all the caches in

`cacheRef.current.filteredOptions`
`cacheRef.current.expandedValues `
`cacheRef.current.selectedValues `

File: js/src/components/tree-select-control/index.js [hunk 2]
```diff
@@ -140,23 +159,29 @@ const TreeSelectControl = ( {
...
 	 */
-	useEffect( () => {
+	const filteredOptions = useMemo( () => {
 		const cachedFilteredOptions = filteredOptionsCache.current[ filter ];
 
+		if ( cachedFilteredOptions ) {
+			return cachedFilteredOptions;
+		}
+
+		const isFiltered = Boolean( filter );
```
Comment (puntope) [inline js/src/components/tree-select-control/index.js]: ❓  Why is this necessary? 

Just using `filter` should work I guess

File: js/src/components/tree-select-control/index.js [hunk 3]
```diff
@@ -169,30 +194,104 @@ const TreeSelectControl = ( {
...
+					const match = option.label.toLowerCase().indexOf( filter );
+					if ( match === -1 ) {
+						return acc;
+					}
+					option.label = highlightOptionLabel( option.label, match );
+				}
+
+				Object.defineProperties( option, descriptors );
+				acc.push( option );
+			}
```
Comment (puntope) [inline js/src/components/tree-select-control/index.js]: 💅 

Just an idea, I guess the reducer could be simplified with: 

```
const descriptors = {
            hasChildren: {
				get() {
					return this.children?.length > 0;
				},
			},
			allLeaves: {
```
```
const reduceOptions = ( acc, { children = [], ...option } ) => {
			if ( children.length ) {
				option.children = children.reduce( reduceOptions, [] );
			} else if ( isFiltered ) {
				const match = option.label.toLowerCase().indexOf( filter );
				if ( match === -1 ) {
					return acc;
				}
				option.label = highlightOptionLabel( option.label, match );
			}

			if ( option.children?.length || ! children.length ) {
				Object.defineProperties( option, descriptors );
				acc.push( option );
			}
			return acc;
		};
```

File: js/src/components/tree-select-control/index.js [hunk 4]
```diff
@@ -46,13 +46,31 @@ import { ARROW_DOWN, ARROW_UP, ENTER, ESCAPE, ROOT_VALUE } from './constants';
...
+
+/**
+ * @typedef {Object} BaseOption
  * @property {string} label The label for the option
  * @property {Option[]} [children] The children Option objects
- * @property {string} [key] Optional unique key for the Option. It will fallback to the value property if not defined
- **/
+ *
+ * @typedef {CommonOption & BaseOption} Option
+ */
```
Comment (puntope) [inline js/src/components/tree-select-control/index.js:60]: 💅  I will use just `Option` instead of CommonOption and BaseOption. (Or maybe I missed the intent of this separation). 

@typedef {Object} Option
@property {string} label The label for the option
@property {string} value The value for the option
@property {string} [key] Optional unique key for the Option. It will fallback to the value property if not defined
@property {Option[]} [children] The children Option objects

💅  WDYT of renaming `InnerOption` as well as `BaseInnerOption` to `Leaf` and `BaseLeaf`

In my mind I see it this way:
Option is the "pure" item we send to the tree, the data itself, with label, value, and optional key and children.
Leaf is and enhanced Option with Tree props (BaseLeaf).

File: js/src/components/tree-select-control/index.js [hunk 5]
```diff
@@ -169,30 +194,104 @@ const TreeSelectControl = ( {
 			);
 		};
 
-		const filterOption = ( option ) => {
-			if ( hasChildren( option ) ) {
-				option.children = option.children.filter( filterOption );
-				return !! option.children.length;
-			}
+		const descriptors = {
+			allLeaves: {
```
Comment (puntope) [inline js/src/components/tree-select-control/index.js]: 💅  I will keep it as `leaves` instead of `allLeaves`. When I saw `all` highlighted in the name of the prop, I was expecting to have just a generic `leaves` somewhere in the code. 

Also the description

`@property {InnerOption[]} allLeaves All (filtered) leaf options under this option.`

I'd suggest to replace it with something like: 

`@property {Leaf[]} leaves All the filtered option children flattened.`

PD. I guess InnerOption is not accurate since we have for example the option.label as well??? 🤔

File: js/src/components/tree-select-control/index.js (hunk 3 above)
Comment (puntope) [reply js/src/components/tree-select-control/index.js]: 💅 WDYT of adding a small comment in the reducer?

Something like "This reducer gets the options matching the filter and also adds Leaf properties to them"

**Review by puntope — COMMENTED** [review]
Howdy @eason9487 

I did a first check for this PR and it works amazing, also it simplifies a lot the relationship between the Options component and the tree ❤️ 

In left some comments tho which I think deserve discussion.

Thanks!

File: js/src/components/tree-select-control/index.js (hunk 1 above)
Comment (eason9487) [reply js/src/components/tree-select-control/index.js]: Changed in https://github.com/woocommerce/google-listings-and-ads/pull/1522/commits/e2a6a827289312f796e202e68b628ed34f0485d4.

File: js/src/components/tree-select-control/index.js (hunk 2 above)
Comment (eason9487) [reply js/src/components/tree-select-control/index.js]: In this context, `isFiltered` is more semantic than `filter`, and the explicit boolean type allows the below return value to always be a boolean type.

https://github.com/woocommerce/google-listings-and-ads/blob/afa8de0035bd66217caa96b5212fb51ff1ea2be9/js/src/components/tree-select-control/index.js#L258-L266

File: js/src/components/tree-select-control/index.js (hunk 3 above)
Comment (eason9487) [reply js/src/components/tree-select-control/index.js]: Thanks for the nice suggestion! Based on your example, I adjusted it to a similar implementation in https://github.com/woocommerce/google-listings-and-ads/pull/1522/commits/da745d8b36ba4e9e020ae00604f85d8fcf238630.

---

> WDYT of adding a small comment in the reducer?

The processing of this reducer is also included in the code comments above, so I guess it's already sufficient?

https://github.com/woocommerce/google-listings-and-ads/blob/afa8de0035bd66217caa96b5212fb51ff1ea2be9/js/src/components/tree-select-control/index.js#L163-L169

File: js/src/components/tree-select-control/index.js (hunk 4 above)
Comment (eason9487) [reply js/src/components/tree-select-control/index.js:60]: > I will use just `Option` instead of CommonOption and BaseOption. (Or maybe I missed the intent of this separation).

Option and InnerOption both have `label` and `children` properties, but the type and description of the properties are different. JSDoc does not override the same attributes when using `@typedef {Option & BaseInnerOption} InnerOption`, but combines them.

For example, using this JSDoc: 

```js
/**
 * @typedef {Object} Option
 * @property {string} label The label for the option.
 * @property {string} value The value for the option.
 * @property {string} [key] Optional unique key for the Option. It will fallback to the value property if not defined.
 * @property {Option[]} [children] The children Option objects.
 */

/**
 * @typedef {Object} BaseInnerOption
 * @property {string|JSX.Element} label The label string or label with highlighted react element for the option.
 * @property {InnerOption[]|undefined} children The children options. The options are filtered if in searching.
 * (omitted)
 * @typedef {Option & BaseInnerOption} InnerOption
 */
```

There will be a combined type and description:



Furthermore, the nested reference produces inconsistent results.




---

> WDYT of renaming `InnerOption` as well as `BaseInnerOption` to `Leaf` and `BaseLeaf`

In a tree data structure, the terminology of *leaf* node refers to the node with no children. But an `InnerOption` could be a parent node, both parent and child node, or a leaf node. Therefore, I don't think `Leaf` is a proper name.

File: js/src/components/tree-select-control/index.js (hunk 5 above)
Comment (eason9487) [reply js/src/components/tree-select-control/index.js]: Renamed the `allLeaves` and adjusted some JSDocs in https://github.com/woocommerce/google-listings-and-ads/pull/1522/commits/afa8de0035bd66217caa96b5212fb51ff1ea2be9.

> PD. I guess InnerOption is not accurate since we have for example the option.label as well???

Sorry, I don't quite understand this. The `InnerOption` does have the `label` property. And what does *PD.* refer to? Could you elaborate more?

File: js/src/components/tree-select-control/index.js (hunk 2 above)
Comment (puntope) [reply js/src/components/tree-select-control/index.js]: When I saw it I thought for a moment that was kind of a flag indicating that a node was filtered or not. 

💅  Maybe consider to change it to `hasFilters`

File: js/src/components/tree-select-control/index.js [hunk 6]
```diff
@@ -169,30 +194,104 @@ const TreeSelectControl = ( {
...
-			if ( hasChildren( option ) ) {
-				option.children = option.children.filter( filterOption );
-				return !! option.children.length;
-			}
+		const descriptors = {
+			hasChildren: {
+				get() {
+					return this.children?.length > 0;
+				},
+			},
```
Comment (puntope) [inline js/src/components/tree-select-control/index.js]: 💅  Missing docs

File: js/src/components/tree-select-control/index.js (hunk 2 above)
Comment (eason9487) [reply js/src/components/tree-select-control/index.js]: > When I saw it I thought for a moment that was kind of a flag indicating that a node was filtered or not.

The naming was taken from the former prop used in the Options component.

https://github.com/woocommerce/google-listings-and-ads/blob/65d232a5aab91dad7910ec352a6389ff4c6f140c/js/src/components/tree-select-control/options.js#L26

> Maybe consider to change it to `hasFilters`

How about `isSearching`?

File: js/src/components/tree-select-control/index.js (hunk 4 above)
Comment (puntope) [reply js/src/components/tree-select-control/index.js:60]: > In a tree data structure, the terminology of leaf node refers to the node with no children. But an InnerOption could be a parent node, both parent and child node, or a leaf node. Therefore, I don't think Leaf is a proper name.

👌  Right. Maybe `node` then. 

> There will be a combined type and description:

Interesting. In my IDE it does the intersection properly.

Screenshot 2022-05-25 at 12 31 34

My worries are that Inner, Base, Common prefixes could be kinda confusing but this is a non blocking issue.

File: js/src/components/tree-select-control/index.js (hunk 5 above)
Comment (puntope) [reply js/src/components/tree-select-control/index.js]: > Sorry, I don't quite understand this. The InnerOption does have the label property. And what does PD. refer to? Could you elaborate more?

Apologies. I missed the label property inside `InnerOption`. All good 👌 (with PD I meant PS, had a typo)

File: js/src/components/tree-select-control/index.js (hunk 4 above)
Comment (eason9487) [reply js/src/components/tree-select-control/index.js:60]: > Maybe `node` then.

`Node` is a more general and abstract naming. When viewing the code like `@param {Node} option`, I don't think it's clearer, and it also seems more irrelevant to the parameter name.

> My worries are that Inner, Base, Common prefixes could be kinda confusing

Sorry, I'm not quite sure how they cause confusion in the scope of these type definitions since the Base and Common prefixes are only used for combining the Option and InnerOption. But I'd like to know if you have other naming ideas to replace them to reduce your confusion.

**Review by puntope — APPROVED** [review]
✅  All the discussion were addressed! This is good to go. Thanks @eason9487 

I left some opinionated comments 💅  that could improve even more the code IMO. Those are optional or could be done in future PR. So this is approved.

File: js/src/components/tree-select-control/index.js (hunk 2 above)
Comment (puntope) [reply js/src/components/tree-select-control/index.js]: isSearching looks pretty accurate for me yes! 🤝

File: js/src/components/tree-select-control/index.js (hunk 4 above)
Comment (puntope) [reply js/src/components/tree-select-control/index.js:60]: All the names I have in mind rename Option to Node for defining the items of the tree and differentiate them from the Options provided by the user.

Option = { OptionValue & OptionData } 
Node = { OptionValue & BaseNode } 

`<Options options="filteredOptions"> to be renamed to <Nodes nodes="filteredNodes">`

But that requires more refactor of the Options component that was named wrongly from scratch by me. (notice in the Options.js component we refer to Nodes in ArrowLeft/ArrowRight docs) 

But this is pretty much a subjective and minor issue (unsure even if it worth it all the refactor in renaming).

File: js/src/components/tree-select-control/index.js (hunk 2 above)
Comment (eason9487) [reply js/src/components/tree-select-control/index.js]: Changed in https://github.com/woocommerce/google-listings-and-ads/pull/1522/commits/6b13282a3297613e3d68cd3589315b9c6721e339.

File: js/src/components/tree-select-control/index.js (hunk 6 above)
Comment (eason9487) [reply js/src/components/tree-select-control/index.js]: Added in https://github.com/woocommerce/google-listings-and-ads/pull/1522/commits/6dfadfc6f50e9938f0f5434a1b4fdad1a9bcf93c.

File: js/src/components/tree-select-control/index.js (hunk 4 above)
Comment (eason9487) [reply js/src/components/tree-select-control/index.js:60]: I think the overall naming adjustments could be a separate PR. Together with, for example, the `value` prop could be internally renamed to `values` or `selectedValues`, and the `nodesExpanded` to `expandedValues`, etc.
